File: app/src/main/java/com/chess/cryptobot/enricher/PairResponseEnricher.java
package com.chess.cryptobot.enricher;

import com.chess.cryptobot.model.Pair;
import com.chess.cryptobot.model.Price;
import com.chess.cryptobot.model.response.OrderBookResponse;
import com.chess.cryptobot.model.response.bittrex.BittrexResponse;
import com.chess.cryptobot.model.response.livecoin.LivecoinOrderBookResponse;

import java.util.List;

public class PairResponseEnricher {
    private Pair pair;
    private final static Double bittrexFee = 0.25;
    private final static Double livecoinFee = 0.18;
    private OrderBookResponse bittrexResponse;
    private OrderBookResponse livecoinResponse;
    private List<Price> bids;
    private List<Price> asks;
    private Double ask;
    private Double bid;
    private boolean bidFromLivecoin;
    private Float minPercent;
    private Double bidQuantity;
    private Double askQuantity;


    public PairResponseEnricher(Pair pair) {
        this.pair = pair;
    }

    public void enrichWithResponse(OrderBookResponse response) {
        if (response instanceof BittrexResponse) {
            bittrexResponse = response;
            pair.setBittrexAsk(response.asks().get(0).getValue());
            pair.setBittrexAskQuantity(response.asks().get(0).getQuantity());
            pair.setBittrexBid(response.bids().get(0).getValue());
            pair.setBittrexBidQuantity(response.bids().get(0).getQuantity());
        } else if (response instanceof LivecoinOrderBookResponse) {
            livecoinResponse = response;
            pair.setLivecoinAsk(response.asks().get(0).getValue());
            pair.setLivecoinAskQuantity(response.asks().get(0).getQuantity());
            pair.setLivecoinBid(response.bids().get(0).getValue());
            pair.setLivecoinBidQuantity(response.bids().get(0).getQuantity());
        }
    }

    public PairResponseEnricher enrichWithMinPercent(Float minPercent) {
        this.minPercent = minPercent;
        Double lAsk = livecoinResponse.asks().get(0).getValue();
        Double lBid = livecoinResponse.bids().get(0).getValue();
        Double bAsk = bittrexResponse.asks().get(0).getValue();
        Double bBid = bittrexResponse.bids().get(0).getValue();

        if ((lBid - bAsk) > (bBid - lAsk)) {
            bid = lBid;
            ask = bAsk;
            bids = livecoinResponse.bids();
            asks = bittrexResponse.asks();
            bidFromLivecoin = true;
        } else {
            bid = bBid;
            ask = lAsk;
            bids = bittrexResponse.bids();
            asks = livecoinResponse.asks();
            bidFromLivecoin = false;
        }

        Pair tmpPair = pair;
        tmpPair.setPercent(0.0f);

        bidQuantity = bids.get(0).getQuantity();
        askQuantity = asks.get(0).getQuantity();

        boolean increaseAsks = (bidQuantity > askQuantity);
        int maxPriceSize = bids.size() > asks.size() ? asks.size() : bids.size();

        for (int i = 1; i < maxPriceSize; i++) {
            if (!enrichedFromStack(tmpPair, i, increaseAsks)) {
                return this;
            }
        }
        return this;
    }

    private boolean enrichedFromStack(Pair tmpPair, int i, boolean increaseAsks) {
        Float percent = countPercent(bid, ask, bidFromLivecoin);
        if (percent > minPercent) {
            updateTempPair(tmpPair);
            tmpPair.setPercent(percent);

            if (increaseAsks) {
                askQuantity = askQuantity + asks.get(i).getQuantity();
                ask = asks.get(i).getValue();
            } else {
                bidQuantity = bidQuantity + bids.get(i).getQuantity();
                bid = bids.get(i).getValue();
            }
            return true;
        } else {
            pair = tmpPair;
            return false;
        }
    }

    private void updateTempPair(Pair tmpPair) {
        if (bidFromLivecoin) {
            tmpPair.setLivecoinBidQuantity(bidQuantity);
            tmpPair.setBittrexAskQuantity(askQuantity);
            tmpPair.setLivecoinBid(bid);
            tmpPair.setBittrexAsk(ask);
        } else {
            tmpPair.setBittrexBid(bidQuantity);
            tmpPair.setLivecoinAskQuantity(askQuantity);
            tmpPair.setBittrexBid(bid);
            tmpPair.setLivecoinAsk(ask);
        }
    }


    private Float countPercent(Double bid, Double ask, boolean bidFromLivecoin) {
        if (bid == null || ask == null) return 0.0f;

        Double bidFee = bidFromLivecoin ? livecoinFee : bittrexFee;
        Double askFee = bidFromLivecoin ? bittrexFee : livecoinFee;
        Float percentBid = Double.valueOf(bid - ((bid / 100) * bidFee)).floatValue();
        Float percentAsk = Double.valueOf(ask + ((ask / 100) * askFee)).floatValue();
        return ((percentBid - percentAsk) / percentBid) * 100;
    }

    public PairResponseEnricher countPercent() {
        if (pair.getLivecoinAsk() == null || pair.getLivecoinBid() == null
                || pair.getBittrexAsk() == null || pair.getBittrexBid() == null
                || pair.getLivecoinAsk() == 0 || pair.getBittrexAsk() == 0) {
            pair.setPercent(0.0f);
            return this;
        }

        Float bb = Double.valueOf(pair.getBittrexBid() - ((pair.getBittrexBid() / 100) * bittrexFee)).floatValue();
        Float ba = Double.valueOf(pair.getBittrexAsk() + ((pair.getBittrexAsk() / 100) * bittrexFee)).floatValue();
        Float lb = Double.valueOf(pair.getLivecoinBid() - ((pair.getLivecoinBid() / 100) * livecoinFee)).floatValue();
        Float la = Double.valueOf(pair.getLivecoinAsk() + ((pair.getLivecoinAsk() / 100) * livecoinFee)).floatValue();

        Float bittrexPercent = ((bb - la) / la) * 100;
        Float livecoinPercent = ((lb - ba) / ba) * 100;
        if (bittrexPercent > livecoinPercent) {
            pair.setPercent(bittrexPercent);
        } else {
            pair.setPercent(livecoinPercent);
        }
        return this;
    }

    public Pair getPair() {
        return pair;
    }
}

File: app/src/main/java/com/chess/cryptobot/enricher/PairResponseEnricher.kt
package com.chess.cryptobot.enricher

import com.chess.cryptobot.market.Market
import com.chess.cryptobot.model.Pair
import com.chess.cryptobot.model.Price
import com.chess.cryptobot.model.response.OrderBookResponse
import com.chess.cryptobot.model.response.TickerResponse
import com.chess.cryptobot.model.response.binance.BinanceResponse
import com.chess.cryptobot.model.response.bittrex.BittrexOrderBook
import com.chess.cryptobot.model.response.livecoin.LivecoinResponse

class PairResponseEnricher(val pair: Pair) {
    private var responsesMap = HashMap<String, OrderBookResponse>(3)
    private var bids: List<Price?>? = null
    private var asks: List<Price?>? = null
    private var bid = 0.0
    private var ask = Double.MAX_VALUE
    private var bidQuantity = 0.0
    private var askQuantity = 0.0
    private var minPercent: Float = 0.0F
    private var feeMap = hashMapOf(
            Market.BITTREX_MARKET to bittrexFee,
            Market.BINANCE_MARKET to binanceFee,
            Market.LIVECOIN_MARKET to livecoinFee)

    fun enrichWithResponse(response: OrderBookResponse) {
        var marketName  = ""
        when (response) {
            is BittrexOrderBook -> {
                marketName = Market.BITTREX_MARKET
            }
            is BinanceResponse -> {
                marketName = Market.BINANCE_MARKET
            }
            is LivecoinResponse -> {
                marketName = Market.LIVECOIN_MARKET
            }
        }
        initMaps(marketName, response)
    }

    fun enrichFromTicker(tickerResponse: TickerResponse, marketName: String): PairResponseEnricher {
        pair.askMap[marketName] = tickerResponse.tickerAsk
        pair.bidMap[marketName] = tickerResponse.tickerBid
        return this
    }

    private fun initMaps(marketName: String, response: OrderBookResponse) {
        responsesMap[marketName] = response
        pair.bidMap[marketName] = response.bids()?.get(0)?.value ?: 0.0
        pair.askMap[marketName] = response.asks()?.get(0)?.value ?: Double.MAX_VALUE
        pair.bidQuantityMap[marketName] = response.bids()?.get(0)?.quantity ?: 0.0
        pair.askQuantityMap[marketName] = response.asks()?.get(0)?.quantity ?: 0.0
    }

    fun enrichWithMinPercent(minPercent: Float?): PairResponseEnricher {
        bid = pair.bidMap.values.maxOrNull() ?:0.0
        ask = pair.askMap.values.minOrNull() ?:Double.MAX_VALUE
        pair.bidMarketName = pair.bidMap.filterValues {it == bid }.keys.first()
        pair.askMarketName = pair.askMap.filterValues { it == ask }.keys.first()
        bidQuantity = pair.bidQuantityMap[pair.bidMarketName]?:0.0
        askQuantity = pair.askQuantityMap[pair.askMarketName]?:0.0
        if (minPercent==null) {
            updatePair()
            pair.percent = countPercent()
            return this
        }
        this.minPercent = minPercent
        bids = responsesMap[pair.bidMarketName]?.bids()
        asks = responsesMap[pair.askMarketName]?.asks()

        val increaseAsks = bidQuantity > askQuantity
        val maxPriceSize = if (bids?.size ?: 0 > asks?.size ?: 0) asks?.size ?: 0 else bids?.size?: 0
        for (i in 1 until maxPriceSize) {
            if (!enrichedFromStack(i, increaseAsks)) {
                return this
            }
        }
        return this
    }

    private fun enrichedFromStack(i: Int, increaseAsks: Boolean): Boolean {
        val percent = countPercent()
        return if (percent > minPercent) {
            updatePair()
            pair.percent = percent
            if (increaseAsks) {
                askQuantity += asks?.get(i)?.quantity ?: 0.0
                ask = asks?.get(i)?.value ?: 0.0
            } else {
                bidQuantity += bids?.get(i)?.quantity ?: 0.0
                bid = bids?.get(i)?.value ?: 0.0
            }
            true
        } else {
            false
        }
    }

    private fun countPercent(): Float {
        val bidFee = feeMap[pair.bidMarketName]?:1.0
        val askFee = feeMap[pair.askMarketName]?:1.0
        val percentBid = (bid - bid / 100 * bidFee).toFloat()
        val percentAsk = (ask + ask / 100 * askFee).toFloat()
        return (percentBid - percentAsk) / percentBid * 100
    }

    private fun updatePair() {
        pair.bidMap[pair.bidMarketName] = bid
        pair.bidQuantityMap[pair.bidMarketName] = bidQuantity
        pair.askMap[pair.askMarketName] = ask
        pair.askQuantityMap[pair.askMarketName] = askQuantity
        pair.bid = bid
        pair.bidQuantity = bidQuantity
        pair.ask = ask
        pair.askQuantity = askQuantity
    }

    companion object {
        private const val bittrexFee = 0.2
        private const val binanceFee = 0.1
        private const val livecoinFee = 0.18
    }
}